compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs

```diff
@@ -1285,14 +1285,36 @@ struct TupleMemberDescriptionFactory<'tcx> {
 
 impl<'tcx> TupleMemberDescriptionFactory<'tcx> {
     fn create_member_descriptions(&self, cx: &CodegenCx<'ll, 'tcx>) -> Vec<MemberDescription<'ll>> {
+        // For closures and generators, name the captured upvars
+        // with the help of `CapturedPlace::to_mangled_name`.
+        let closure_def_id = match *self.ty.kind() {
+            ty::Generator(def_id, ..) => def_id.as_local(),
+            ty::Closure(def_id, ..) => def_id.as_local(),
+            _ => None,
+        };
+        let captures = match closure_def_id {
+            Some(local_def_id) => {
+                let typeck_results = cx.tcx.typeck(local_def_id);
+                let captures = typeck_results
+                    .closure_min_captures_flattened(local_def_id.to_def_id())
+                    .collect::<Vec<_>>();
+                Some(captures)
+            }
+            _ => None,
+        };
```

Comment:
I don't understand why this is in `TupleMemberDescriptionFactory`, but it's probably my fault.

Is there a simple refactor that would remove the special casing from this path? It seems like reusing the tuple code for closures was useful at one time, but now it seems awkward.

If not don't worry about it.

Reply:
Do you mean to build a new `MemberDescriptionFactory` for captured variables? I am able to do this, but it may be not necessary. The reason is that we will fallback to tuples if closures are not local (i.e. closures in external crates that are compiled in because of monomorphization).

Reply:
Do you mean to build a new `MemberDescriptionFactory` for captured variables? I am able to do this, but it may be not necessary. The reason is that we will fallback to tuples if closures are not local (i.e. closures in external crates that are compiled in because of monomorphization).

Reply:
Actually I found this issue just now. Then I have checked `rustc_metadata`. I seems that we don't store the information about captures now. So I have to add a new query and amend `rustc_metadata` in the fix. I can try to do this, either in this PR or in a new PR.

Reply:
Oh I've found this:
https://github.com/rust-lang/rust/blob/75da570d784a798a34ff1e5048cd9a6a2fb23170/compiler/rustc_mir_build/src/build/mod.rs#L926-L950
So perhaps I can query `optimized_mir` instead of `typeck_results` to fetch the name.